Source/WebKit/Platform/IPC/ThreadSafeObjectHeap.h

```diff
@@ -91,7 +97,45 @@ class ThreadSafeObjectHeap {
 };
 
 template<typename Identifier, typename HeldType>
-HeldType ThreadSafeObjectHeap<Identifier, HeldType>::retire(ReadReference&& read, Timeout timeout)
+bool ThreadSafeObjectHeap<Identifier, HeldType>::add(const Reference& ref, HeldType object)
+{
+    Locker locker { m_objectsLock };
+    auto addResult = m_objects.ensure(ref, [&object]() mutable {
+        return ReferenceState { WTFMove(object) };
+    });
+    if (!addResult.isNewEntry) {
+        auto& state = addResult.iterator->value;
+        if (state.finalRead && !*(state.finalRead)) {
+            // `object` will be destroyed without the lock later.
+            m_objects.remove(addResult.iterator);
+            return true; // Reference was already removed, so nobody is waiting. No need to notify.
+        }
+        if (state.object)
```

Comment:
Added some HashTrait usage so that Ref works.
Note: The ..State::object is  optional<T>, if(state.object) works. 
Note: Ref is not move-only, it has copy operation. Ref is non-nullable.

Changed the tests to use Ref so it's more apparent.